### PR TITLE
Add `Karras sigmas` to HeunDiscreteScheduler

### PR DESCRIPTION
**Pull request by youssefadr**
Discussion: #2905 #2064 

This pull request allows for the utilization of karras_sigmas in the Heun Discrete scheduler and builds upon the valuable contributions made in PR #2956. This update introduces two key components:

- The inclusion of karras-based sigma.
- The mapping of sigmas to the timestamp, which will be used as input for UNET.

**Heun Discrete Test**

```python
from diffusers import StableDiffusionPipeline, HeunDiscreteScheduler
import torch 

seed = 33

pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cpu")
pipe.scheduler = HeunDiscreteScheduler.from_config(pipe.scheduler.config, use_karras_sigmas=True)

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cpu").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]

display(image)
```

without_k

**Compared with KDiffusion - sample_heun**

```python
import torch
from diffusers import StableDiffusionKDiffusionPipeline

seed = 33

pipe = StableDiffusionKDiffusionPipeline.from_pretrained('CompVis/stable-diffusion-v1-4').to('cpu')
pipe.set_scheduler('sample_heun')

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cpu").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20, use_karras_sigmas=True).images[0]

display(image)
```

with_k

### CODE REVIEW

**Comment by youssefadr** [comment]
Hello @sayakpaul and @patrickvonplaten,

Could you please review this PR? 
Much appreciated!

**Review by sayakpaul — APPROVED** [review]
Well done 🔥 

Good to go for me as soon as the CI is green.

**Comment by youssefadr** [comment]
Thank you for your feedback @sayakpaul. The CI should be ok now!

**Comment by youssefadr** [comment]
I've made updates to the pull request, including the fix introduced in commit `1747005`, for the previously failing code quality check. 
However, there are now failing CI tests associated with UNet3DConditionModelTests.

**Comment by sayakpaul** [comment]
The test failure is unrelated now. 

@patrickvonplaten seems like the tests for `UNet3DConditionModel` are again becoming a bit flaky.

**Comment by patrickvonplaten** [comment]
Great job! We can merge this IMO
